torch/csrc/jit/codegen/cuda/lower_unroll.cpp

```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+    }
+  }
+  return loops;
+}
+
+// Find the first vectorize set - either read or write
+// Add child For-Loop to loop_structure
+// Enable vectorize flag in child For-Loop
+kir::Expr* findVectorizedSet(
+    std::vector<kir::ForLoop*>& loop_structure,
+    const std::vector<kir::ForLoop*>& for_loops) {
+  for (auto fl : for_loops) {
+    auto first_expr = fl->body().exprs().front();
+    bool has_vectorize_op =
+        (first_expr->isA<kir::UnaryOp>() &&
+         first_expr->as<kir::UnaryOp>()->operation() == UnaryOpType::Set &&
+         fl->iter_domain()->parallelType() ==
+             ParallelType::MisalignedVectorize);
+    if (has_vectorize_op) {
+      loop_structure.push_back(fl);
+      return first_expr;
+    }
+  }
+  return nullptr;
+}
+
+// Get full extent for the inner-most, merged root domain
+kir::Val* getVectorizeExtent(
+    kir::TensorView* tv,
+    const std::vector<kir::Val*>& indices) {
+  auto domain = tv->domain()->hasRFactor() ? tv->domain()->rfactorDomain()
+                                           : tv->domain()->rootDomain();
+
+  TORCH_INTERNAL_ASSERT(domain.size() == indices.size());
+
+  bool is_contiguous = true;
+  for (auto status : tv->domain()->contiguity()) {
+    is_contiguous &= status;
+  }
+
+  // If the tensorview is not contiguous, return inner-most root domain extent
+  if (!is_contiguous) {
+    return domain.back()->extent();
+  }
+
+  const auto gpu_lower = GpuLower::current();
+  kir::IrBuilder ir_builder(gpu_lower->kernel());
+
+  // Calculate extent of merged root domains
+  kir::Val* extent = nullptr;
+  for (int i = int(domain.size()) - 1; i >= 0; --i) {
+    auto root_id = domain.at(i);
+    if (root_id->isBroadcast() || root_id->isReduction() ||
+        gpu_lower->trivialReductionInfo().isDerived(root_id)) {
+      continue;
+    } else if (extent == nullptr) {
+      extent = root_id->extent();
+    } else if (extent != nullptr && indices.at(i)->isZeroInt()) {
+      // This root id must be merged and contiguous. Expand the
+      // vectorization partition.
+      extent = ir_builder.mulExpr(extent, root_id->extent());
+    } else {
+      break;
+    }
+  }
+
+  TORCH_INTERNAL_ASSERT(extent != nullptr);
+
+  return extent;
+}
+
+kir::Val* setupNamedScalar(
+    kir::Scope& body,
+    kir::Val* val,
+    const std::string& name,
+    bool address = false) {
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+  auto namedScalar = (address) ? ir_builder.namedAddressExpr(name, val)
+                               : ir_builder.namedSetExpr(name, val);
+  auto alloc = ir_builder.create<kir::Allocate>(
+      namedScalar, MemoryType::Local, ir_builder.one());
+  body.push_back(alloc);
+  body.push_back(namedScalar->definition());
+  return namedScalar;
+}
+
+kir::ForLoop* handleMisalignedVectorization(
+    std::vector<kir::ForLoop*> loop_structure,
+    const kir::ForLoop* for_loop) {
+  // for_loop body contains allocate, read, compute, write operations
```

Comment:
Why? What is it?

Reply:
The base for-loop contains the for-loops with the vectorized sets.
```
// base for-loop 
for (....) {
  // Allocations
  for (...) {
    // Read - UnaryOp::Set - Vectorize
  }
  for (...) {
     // Computation
  }
  for (...) {
    // Write - UnaryOp::Set - Vectorize
  }
}
```